Feat/progress events

jspizziri opened the pull request.

jspizziri commented:
@dcvz this PR is at a point where some feedback would be great. It seems to be working on Android/iOS the way I would expect, but I'm not a native expert.

Also, remind me again why it wouldn't be a good idea to update the `useProgress` hook to use this? It seems to me that it would be more efficient as it's only a single native bridge execution rather than 3.

martinmidtsund commented:
@jspizziri Just chiming in on your question, at least for us, the interval will be shorter for the useProgress hook when the app is in the foreground. And we want this event to run at a longer interval at the same time, i.e. both foreground and background. Can this work when using this for the useProgress hook?

jspizziri commented:
@martinmidtsund so the way it exists now it couldn't. However, I think it could probably be expanded to handle both scenarios (both long & short intervals).

martinmidtsund commented:
@jspizziri That would be great. What did you have in mind? Two different event names, configured separately?

jspizziri commented:
@martinmidtsund here are the options I can think of:

1. A separate event (which isn't ideal since you're sending the same exact data over the bridge 2x)
2. Debouncing (which David said isn't possible, and I believe him)
3. Have users manually debounce.
    - So this is less ideal from users perspective, but you could effectively keep a counter of how many times an event is called and use that to gate your processing (see example below)
    
```ts
let eventCount = 0;

TrackPlayer.addEventListener(Event.PlaybackProgressUpdated, (payload) => {
    eventCount++;
    
    // if this event was called on a 1s interval then this would gate an action to roughly 5s
    if (eventCount % 5 === 0) {
      // do stuff.
    }
});
```

dcvz commented on android/src/main/java/com/doublesymmetry/trackplayer/service/MusicService.kt:
```diff
@@ -176,10 +179,46 @@ class MusicService : HeadlessJsTaskService() {
+        val updateInterval = Utils.getIntOrNull(options, PROGRESS_UPDATE_EVENT_INTERVAL_KEY)
+        if (updateInterval != null && updateInterval > 0) {
+            val updateIntervalMs = (updateInterval * 1000).toLong()
+            progressUpdateRunnable = Runnable {
```
Exoplayer does not have a progress updated callback, so usually using a Runner is the right approach but we're using Kotlin coroutines in this file so we'll want to us that instead. You can create a "timer" using flow.

jspizziri replied:
@dcvz updated this to a flow, let me know if it looks correct.

dcvz reviewed (changes requested):
Some initial comments

dcvz commented:
@jspizziri my thoughts on why we can't use this for the progress hook was:

1. We don't know if and what interval the users will configure it as.
_the progress hook on the other hand is typically used for progress indicators and we want that update to be very regular (every second) or so.. using this also for the hook then becomes more complicated_

2. I thought powering the progress hook via this event would be heavy on the JS bridge since because of point 1 you'd need this to happen at least every second.. therefore sending updates over the bridge. And currently the progress hook uses a timer on the JS side -- but the hook requests three properties from the native side, which travel from through the bridge anyway...

But overall I think point 1 is a big one which is not easy to overcome and I don't think we should make two events.

jspizziri commented on ios/RNTrackPlayer/RNTrackPlayer.swift:
```diff
@@ -316,9 +318,47 @@ public class RNTrackPlayer: RCTEventEmitter, AudioSessionControllerDelegate {
+
+        guard let interval = progressUpdateEventInterval else { return }
+
+        DispatchQueue.main.async {
```
@dcvz do I need to extend the `QueuedAudioPlayer` in order to implement that method? Sorry for the n00bery.

dcvz replied:
No. So we already create a `QueuedAudioPlayer` and on the init of this file you see we subscribe to events. There's an event for `secondElapse` that you can subscribe to which fires based on the configured interval.

jspizziri replied:
Interesting, I had tried that but no events were firing. I'll try again. Do I need to subscribe _after_ setting the time elapsed? Or can I do it globally with all the other listener subscriptions.

jspizziri replied:
Nvm, I got it working now. Thanks!

jspizziri replied:
Alright, one issue is that there doesn't seem to be a way to _unset_ `timeEventFrequency`.  Moreover, its value is defaulted to `1s`. I've tried things like setting it to values like  `nil`, `TimeEventFrequency.custom(time: CMTime.indefinite)` (also `negativeInfinity`, `invalid`, `zero`), and even `0s`.

All of which result in an exception:

```
invalid parameter not satisfying: ((Boolean)(((interval).flags & (kCMTimeFlags_Valid | kCMTimeFlags_ImpliedValueFlagsMask)) == kCMTimeFlags_Valid))' was thrown
```

jspizziri replied:
I suppose for now I can just set a class level property and check that property each time the event is emitted, but it seems less than ideal.

jspizziri commented:
@dcvz I've updated the PR to the best of my current knowledge and it's ready for you to look at again when you have a chance.

dcvz commented on android/src/main/java/com/doublesymmetry/trackplayer/service/MusicService.kt:
```diff
@@ -19,16 +19,18 @@ import com.doublesymmetry.trackplayer.utils.Utils.setRating
 import com.facebook.react.HeadlessJsTaskService
 import com.facebook.react.bridge.Arguments
 import com.facebook.react.bridge.Promise
+import com.facebook.react.bridge.WritableMap
```
some changes regarding android but I offer them as a patch instead:
[0001-Android-changes.patch.zip](https://github.com/doublesymmetry/react-native-track-player/files/8535075/0001-Android-changes.patch.zip)

dcvz replied:
@jspizziri according to the docs, we should actually not use this experimental `Duration` object, so let's stick to the longs you had before in my patch

dcvz reviewed (commented):
Some more code improvement comments

jspizziri commented on ios/RNTrackPlayer/RNTrackPlayer.swift:
```diff
@@ -821,4 +838,21 @@ public class RNTrackPlayer: RCTEventEmitter, AudioSessionControllerDelegate {
+                "position": player.currentTime,
+                "duration": player.duration,
+                "buffered": player.bufferedPosition,
+                "track": track.toObject(),
```
@dcvz I figured we'd have a discussion about this. My rationale here was that we should pass back some data which identified the track the progress update referred to. If it's not passed then the first thing everyone is going to have to do is identify the track somehow. The most convenient way from my perspective is to pass the Track object as people likely have some sort of ID on it (even though it's not required, but allowed for based on the Track interface). The other option would be to pass back the index of the track which is probably workable but at least from my perspective isn't as convenient.

dcvz replied:
Yeah.. I guess this event won't be so helpful without the context of the track you're getting an update for.

_What would be the alternative?_
You'd have to make a call to `getCurrentTrack()` to get the index and `getTrack()` to get the object.

Is this so bad though?

-----

The two big use cases were:
1. Sleep timers (for this you probably don't really need the current track info)
2. Analytics (people might want the current track data?)

My biggest concern is just that this `track` object could be any payload the user sent in as the track.. so we're potentially sending variable size objects that the user may or may not want.

jspizziri replied:
My spidey-sense was tingling for that same reason. How about this:

If a _track.id_ is set we send it alone back in the payload, if not we send nothing.If people care about it then they can set an id on their track. It's more opinionated but maybe that's a good thing and it's a fixed size.

The thing I worry about with fetching data based on indexes in JS is latent events towards the end of a track playback. You could end up setting an incorrect progress for a track based on timing without a concrete way to ID the event.

jspizziri replied:
@dcvz want to make sure this is showing up for you (since we've been going back and forth today)

dcvz replied:
This might be a good first step! We may have some time during the RC's to have others who need it test it and give feedback. I'm ok with proceeding with this

dcvz replied:
I guess we're tailoring the implementation too much towards knowing that people will use of for analytics as opposed to sending the track index and allowing them to query for the track object if they want.

jspizziri replied:
So are you saying you'd rather send an `index` instead of a `trackId`? I suppose it's relatively unlikely that the index will change during an even processing.

dcvz replied:
Yeah.. thinking more about it I think we should return the index. Everything in this library is index based, so I think that makes sense.

Users can still grab the track object if need be by calling `getTrack`. Returning an identifier is too tailored towards analytics and forces people to use an `id` if they want to take advantage of this.

Now they can use any keys they want and only get the object if they want to for their purposes.

jspizziri replied:
I'm fine with that since it seems it would be highly unlikely for the position to change. However, if there are valid cases where the position in the queue can change during these events (for instance someone inserting at a specific index, which isn't something I do) it may be the case that we _also_ need to return a `trackId` too.

jspizziri commented:
@dcvz implemented your feedback and left you my rationale on emitting `track` as part of the event payload.

dcvz commented on ios/RNTrackPlayer/RNTrackPlayer.swift:
```diff
@@ -316,9 +319,21 @@ public class RNTrackPlayer: RCTEventEmitter, AudioSessionControllerDelegate {
                                           bookmarkOptions: options["bookmarkOptions"] as? [String: Any])
         }
 
+        if let interval = options["progressUpdateEventInterval"] as? NSNumber, Int(truncating: interval) > 0 {
```
```suggestion
        if let interval = options["progressUpdateEventInterval"] as? NSNumber, interval.intValue > 0 {
```

I think this would work too..

dcvz reviewed (commented):
Looks good on my side -- still pending discussion on returning the track object.
But need to think about it more and open to your opinions.

I'd like for Milen to take a look at the Android side before we merge it though :)

jspizziri commented:
@dcvz , I've:

1. Updated the event payload to contain a `trackIndex` instead of a `track` object.
    - I know on all the events it's just `track`, but I think being verbose (especially here) is better. LMK what you think.
2. Rebased on `feature/kotlinaudio` and moved the feature docs to their new homes.
3. Fixed an incorrect URL in the `hooks.ts` file I noticed.
4. Updated the `intValue` you mentioned.
5. Removed a redundant `Event.*` prefix from the event section headings on the events page (since I was in there). The beginning of the doc mentions they're all available on the `{ Event }` enum so having them there I think is more of a distraction.

At the very least I'll want to run one final test once I'm back in the office before this gets merged (since I rebased and the changes to code I made today I didn't test).

jspizziri commented:
I split out the versioning to https://github.com/doublesymmetry/react-native-track-player/pull/1495 which is based off of this branch.

jspizziri commented:
alright, everthing here should be properly ready for review now.

dcvz commented:
@jspizziri since we merged `kotlinaudio` we'll need to reopen this PR. Sorry!

jspizziri commented:
Reopened https://github.com/doublesymmetry/react-native-track-player/pull/1499
